### ISSUE TITLE
Does Buck support Scala?

### ISSUE THREAD

**Issue by Nickersoft** [facebook/buck#1721]
Looking through the source code, I noticed several references to `scala_library()`, and using it in my BUCK file seems to work (except it asks me to specify a Scala library dependency). However, there does not seem to be any reference to Scala at all on the documentation site. Does this library support Scala or am I missing something?

**Comment by kageiit** [comment]
buck supports scala, but you do have to set

```
[scala]
    compiler = ...
    library = ...
```

in your .buckconfig to get the scala rules to compile

**Comment by Nickersoft** [comment]
@kageiit Is there any particular reason that it's not mentioned in the documentation? Should I add it?

Also, I had previously tried this... and had `library` pointed to a BUCK file which attempted to download the Scala library via `remote_file()` and the URL `org.scala-lang:scala-library-all:pom:2.12.4`, however, it errored and said it couldn't download it with no explanation. I've since deleted that code, but can try it again in order to provide more detail about my setup and the error.

**Comment by kageiit** [comment]
Feel free to make a PR to add the docs. Would be helpful if you post the error so we can debug

**Comment by Nickersoft** [comment]
@kageiit Alright I got a quick setup going with the same error. Here's my BUCK file:

```python
remote_file(
    name = 'scala-compiler-source',
    out = 'scala-compiler-source.jar',
    url = 'org.scala-lang:scala-compiler:jar:2.12.4',
    sha1 = '557edd918fd41f9260963583ebf5a61a43a6b423',
)

prebuilt_jar(
    name = 'scala-compiler-lib',
    binary_jar = ':scala-compiler-source',
    visibility = ['//:rake']
)

java_binary(
    name='scala-compiler',
    main_class='scala.tools.nsc.Main',
    deps=[':scala-compiler-lib'],
)

remote_file(
    name = 'scala-library-source',
    out = 'scala-library-source.jar',
    url = 'org.scala-lang:scala-library-all:pom:2.12.4',
    sha1 = '557edd918fd41f9260963583ebf5a61a43a6b423',
)

prebuilt_jar(
    name = 'scala-library',
    binary_jar = ':scala-library-source',
    visibility = ['//:rake']
)

scala_library(
    name = "rake",
    source = "1.8",
    target = "1.8",
    srcs = glob([
        "src/**/*.java"
    ])
)
```

which was based loosely off of some of the source code I found in this repo. My .buckconfig:

```config
[httpserver]
    port = 0

[download]
    in_build = true
    maven_repo = http://central.maven.org/maven2/

[scala]
    compiler = //:scala-compiler
    library = //:scala-library
```

and the output I get running `buck build //:rake`:

```
Unable to download: org.scala-lang:scala-library-all:pom:2.12.4
Unable to download: org.scala-lang:scala-compiler:jar:2.12.4
Build failed: Command failed with exit code -1.
    When running <curl>.
    When building rule //:scala-library-source.
```

[and here's a gist of the full output when the verbosity level is set to 8.](https://gist.github.com/Nickersoft/d97ba35db7504692ca2fa994f3e4f908)

Hopefully this is useful.. skimming through the output the errors seem pretty generic, but I could be overlooking something. Also, I'll start looking into the contributing guidelines and how to update the docs. The Scala code in this repo is from two years ago so it's crazy it never got added.

**Comment by kageiit** [comment]
@Nickersoft the remote file url seems incorrect

according to https://buckbuild.com/rule/remote_file.html , you need a `mvn:` prefix before your maven coordinates

**Comment by Nickersoft** [comment]
@kageiit Oh wow, my bad. I forgot to include the `mvn:` prefix this time around, but in my original code I had it in and still got an error. I added the prefix back to my URLs and updating the hashes to rule out hash errors I still get an error:

```
Unable to download: mvn:org.scala-lang:scala-library-all:pom:2.12.4
Build failed: Command failed with exit code -1.
    When running <curl>.
    When building rule //:scala-library-source.
```

[And like before, here's the v8 output](https://gist.github.com/Nickersoft/06e09fbf2e17ea9166b6124623f3fa80). 

It would appear it's checking for the existence of `http://central.maven.org/maven2/org/scala-lang/scala-library-all/2.12.4/scala-library-all-2.12.4-pom.jar`, which doesn't exist. However, `http://central.maven.org/maven2/org/scala-lang/scala-library-all/2.12.4/scala-library-all-2.12.4.pom` *does* exist, which Gradle still seems to be able to understand despite Buck failing to.

I actually just did a check and changed my task to:

```python
remote_file(
    name = 'scala-library-source',
    out = 'scala-library-source.jar',
    url = 'mvn:org.scala-lang:scala-library:jar:2.12.4',
    sha1 = '7663f74ef944453c86cc7e6597ed33e9281f6412',
)
```

and it seems to work now... I'm not super familiar with how pom files are handled, so I guess I'll just ignore it. Anyway, I'll close this for now seeing it seems to be working and I'll look into updating the docs.
